### DISCUSSION TITLE
Fixed point physics

### DISCUSSION THREAD

**Discussion by richardhozak** [dimforge/rapier#482] [Q&A]
Hello, are there any plans to eventually implement fixed point physics simulation for guaranteed cross platform determinism in addition to f32 and f64?

I know that rapier has "enhanced determinism" feature which can be enabled to achieve cross platform floating point determinism, but I found that it is quite easy to introduce non-determinism by using floats before passing them to rapier.

I remember that nphysics allowed fixed point numbers to be used for simulation and achieved determinism through that, which is quite unique, as I cannot find any physics engine that would allow the use of fixed point numbers, so I was wondering if it will be reintroduced in some way shape or form. 

Also, I would like to ask if you would recommend using nphysics if fixed point physics is required given it is passively maintained and being superseded by rapier?

Thank you in advance for answering my questions and thank you for all your hard work.

**Comment by richardhozak** [comment]
So, I've decided to put my mouth where my money is and try to implement this myself. 

During the implementation I found out that `simba` provides partial support for fixed point numbers, but not enough to be used in `parry` which is needed for `rapier` so I implemented the missing functionality which can be found here https://github.com/dimforge/simba/pull/48 and here is the integration into `parry` https://github.com/dimforge/parry/pull/154 (`parry` also contained some code paths for fixed point support) next thing would be to use this version of `parry` in rapier so it can implement fixed point physics.

**Comment by NailxSharipov** [comment]
3 years ago I did a small prototype in swift with fexed point physics. Probably it can be interesting.
[project](https://github.com/iShape-Swift/iFixBox) and small [demo](https://github.com/iShape-Swift/iFixBox/issues/1)
